### PR TITLE
RFC: arch: arm: Remove aarch32 directory and split Cortex-M and Cortex-A/r code

### PR DESCRIPTION
**Pull request by SgrrZhf**
This PR tries to do some cleaning work in the `arch/arm/core` directory and split some common code into Cortex-M and Cortex-A/R parts. This is a huge work, hope you can give your comments on this.

Below is the major changes:

1. It doesn't make sense to keep the aarch32 directory in the 'arch/arm/core' directory as the aarch64 has been moved out. This PR try to clean the 'aarch32' directory in 'arch/arm/core' and 'include/zephyr/arch/arm'.

1. There are too many differences between Cortex-M and Cortex-A/R, For easier to maintain, this PR try to split some common code in 'arch/arm/core'. They are:
    - arch/arm/core/swap.c
    - arch/arm/core/swap_helper.S
    - arch/arm/core/irq_manage.c
    - arch/arm/core/isr_wrapper.S

1. Move 'arch/arm/core/irq_relay.S to the cortex_m directory because it is only used by Cortex-M architecture.

There is a known issue about cleaning the 'aarch32' directory in the `include/zephyr/arch/arm` directory that `boot/zephyr/arm_cleanup.c`  in the mcuboot module includes a header file resided in `include/zephyr/arch/arm/aarch32` directory. It will lead to a build error.

https://github.com/zephyrproject-rtos/mcuboot/blob/74c4d1c52fd51d07904b27a7aa9b2303e896a4e3/boot/zephyr/arm_cleanup.c#L7C43-L7C52

### CODE REVIEW

**Comment by carlescufi** [comment]
@ibirnbaum please take a look, since you have some unmerged related work.

**Comment by microbuilder** [comment]
Thanks for putting this (non-trivial) PR together. Separating A/R and M has been something we've discussed for a while, but given the scope hadn't happened so far.

@carlescufi Perhaps we should get this on the Arch call in a week, giving people (including myself) some time to dig into it this week?

**Comment by SgrrZhf** [comment]
I just add @povergoing into the reviewers list and I don't know why some people are removed from it. 😥

**Comment by povergoing** [comment]
> @ibirnbaum please take a look, since you have some unmerged related work.

I guess you mean this one https://github.com/zephyrproject-rtos/zephyr/pull/41520

**Comment by ibirnbaum** [comment]
Hey, someone volunteered to tackle 'the big one' :-) Thanks @SgrrZhf ! I'll look into the suggested modifications ASAP, which is likely going to be sometime next week.

@carlescufi @povergoing

I thought about reviving #41520, but so far haven't gotten around to it, as the changes proposed passed the kernel testsuite, so there isn't that much left to do. By the time this PR went stale, there had been comments regarding performing a certain operation identically to how it's done in arm64, which I didn't understand right away. When this PR was already closed, I eventually got a chance to look at the arm64 code and understood that the subject of that comment wasn't actually that complex, but again, I had to postpone my work and move on to other things. Basically, what it was all about was to make sure that at one particular point, a certain value is only read but not overwritten, as a write operation should only be performed once during a context switch operation. As far as I can tell, I was probably 95% there, although in some spots the solution didn't look that pretty (switching CPU modes in order to retrieve certain thread-related values from the stack of another mode etc.). IIRC, a very specific order of providing thread information to the context switch routine had to be observed, otherwise, the switch out  of the initial dummy thread into the first 'proper' thread didn't work.

From my point of view, it doesn't really matter if we re-open the ARCH_SWITCH PR, fix the last few details & hopefully merge it before this PR takes all the so far shared A/M/R sources apart, or if we separate the sources first & then file a new PR inserting the same few lines of code at the same spots in the cleaned up source files. Which one of the two approaches would you prefer? I'll try my best to eventually merge the ARCH_SWITCH support as a necessary pre-condition for SMP. It'd be great to have SMP, it would come in handy to me both on the Zynq (dual A9) and the R-Cores of the UltraScale (dual R5).

Then, there's also still #53262 pending, where I provided input, but I'm not the original author. This PR modifies the linker command file in the aarch32 include directory and should therefore be affected if directories are to be renamed. We should be able to merge this quickly, the suggested changes would fix the issue of unnecessarily large binaries because of bad placement of certain sections within the binary. In case the original author is no longer active, I could open a PR with the last missing change being added on top of what's proposed in #53262.

**Comment by SgrrZhf** [comment]
Hi @ibirnbaum

Please open the ARCH_SWITCH PR first because we need it for enabling SMP for Cortex-A/R aarch32. Also, we are enabling SMP on Cortex-A/R and it would be finished in the next month if it's going well, So we can help take over this as @jaxson said if you are too busy. ^_^

**Comment by microbuilder** [comment]
@ibirnbaum feel free to raise a new PR for https://github.com/zephyrproject-rtos/zephyr/pull/53262 and assign me and I'm happy to review. With the original author gone silent, a new one will likely be quicker to get in with the required sign off.

**Comment by SgrrZhf** [comment]
I noticed that there is a PR(https://github.com/zephyrproject-rtos/mcuboot/pull/106) that tries to replace `zephyr/arch/arm/aarch32/cortex_a_r/cmsis.h` with `cmsis_core.h`.

This is also needed by this PR, so may I ask that if there is any plan for these two PRs? @gmarull 

https://github.com/zephyrproject-rtos/mcuboot/pull/106
https://github.com/zephyrproject-rtos/zephyr/pull/59855

**Comment by gmarull** [comment]
> I noticed that there is a PR([zephyrproject-rtos/mcuboot#106](https://github.com/zephyrproject-rtos/mcuboot/pull/106)) that tries to replace `zephyr/arch/arm/aarch32/cortex_a_r/cmsis.h` with `cmsis_core.h`.
> 
> This is also needed by this PR, so may I ask that if there is any plan for these two PRs? @gmarull
> 
> [zephyrproject-rtos/mcuboot#106](https://github.com/zephyrproject-rtos/mcuboot/pull/106) #59855

If you're happy with the approach I can rebase the PRs

**Comment by SgrrZhf** [comment]
As the https://github.com/zephyrproject-rtos/zephyr/pull/53262 is merged, I think I can move on.
I want to create a new PR to fit this PR for https://github.com/zephyrproject-rtos/mcuboot and rebase it, after that, the CI should all be passed. Are there other things that I should do in this PR? @microbuilder .

**Comment by microbuilder** [comment]
> As the #53262 is merged, I think I can move on. I want to create a new PR to fit this PR for https://github.com/zephyrproject-rtos/mcuboot and rebase it, after that, the CI should all be passed. Are there other things that I should do in this PR? @microbuilder .

You can raise an MCUBoot PR in parallel, and then update the west manifest in this PR to point to the MCUBoot PR (set the revision field to something like `revision: pull/54/head` if it was MCUBoot PR 54). That should enable CI to include the MCUBoot changes here as well, and we'll get a DNM label until the MCUBoot PR is merged.

**Comment by SgrrZhf** [comment]
As the https://github.com/zephyrproject-rtos/zephyr/pull/59855 is merged, I don't need to do with the mcuboot repo anymore.

**Comment by SgrrZhf** [comment]
Hi @microbuilder @carlocaione.
I think this PR is ready to be merged, so is there anything that I need to change to this PR from your opinions.

File: soc/arm/nxp_lpc/lpc11u6x/soc.c
Comment (carlocaione) [inline soc/arm/nxp_lpc/lpc11u6x/soc.c:34]: why is this here and how this is related to this commit?

**Review by carlocaione — CHANGES_REQUESTED** [review]

File: soc/arm/nxp_lpc/lpc11u6x/soc.c
Comment (SgrrZhf) [reply soc/arm/nxp_lpc/lpc11u6x/soc.c:34]: This must be mistake occurred in rebase. I'll check them again.

Comment (SgrrZhf) [reply soc/arm/nxp_lpc/lpc11u6x/soc.c]: I just checked them again so this time should be ok.

**Review by carlocaione — DISMISSED** [review]
Not easy to review but nothing stands out as wrong

**Review by povergoing — DISMISSED** [review]

**Comment by ibirnbaum** [comment]
Wow, the low-level assembly sources are definitely becoming a lot easier to read!
However, I'm a bit surprised to find references to the Cortex-M's BASEPRI register in A/R code (arch/arm/core/cortex_a_r/swap.c) 🤔 Does this actually compile, or is this file being skipped at the time being due to ARCH_SWITCH support not being there yet?

**Comment by SgrrZhf** [comment]
> Does this actually compile, or is this file being skipped at the time being due to ARCH_SWITCH support not being there yet?

It looks like this will be compiled on Cortex-A/R. 

https://github.com/zephyrproject-rtos/zephyr/blob/1ddabfa8d8a6d807717dcd4ce0d6a5ee723a89fb/arch/arm/core/aarch32/swap_helper.S#L271-L273

This part is shared by Cortex-M and Cortex-A/R currently, so I put it into Cortex-A/R in this PR. I'm not sure about that whether Cortex-A/R needs BASEPRI-related code.

**Comment by ibirnbaum** [comment]
```
int arch_swap(unsigned int key)
{
	/* store off key and return value */
	_current->arch.basepri = key;
	_current->arch.swap_return_value = _k_neg_eagain;

	z_arm_cortex_r_svc();
	irq_unlock(key);

	/* Context switch is performed here. Returning implies the
	 * thread has been context-switched-in again.
	 */
	return _current->arch.swap_return_value;
}
```

This is the code from swap.c I was referring to. As far as I can tell, the _current->arch.basepri assignment would only work on a Cortex-M.

**Comment by SgrrZhf** [comment]
> This is the code from swap.c I was referring to. As far as I can tell, the _current->arch.basepri assignment would only work on a Cortex-M.

Got it. However, I would like to keep them in this PR because this time I want to focus on separating them rather than fixing issues. If the _current->arch.basepri assignment is useless on Cortex-A/R, maybe I can create another PR to clean it later.

**Comment by povergoing** [comment]
> ```
> int arch_swap(unsigned int key)
> {
> 	/* store off key and return value */
> 	_current->arch.basepri = key;
> 	_current->arch.swap_return_value = _k_neg_eagain;
> 
> 	z_arm_cortex_r_svc();
> 	irq_unlock(key);
> 
> 	/* Context switch is performed here. Returning implies the
> 	 * thread has been context-switched-in again.
> 	 */
> 	return _current->arch.swap_return_value;
> }
> ```
> 
> This is the code from swap.c I was referring to. As far as I can tell, the _current->arch.basepri assignment would only work on a Cortex-M.

Or maybe... we might totally drop the `arch_swap` and use `USE_SWITCH` to support A/R to support A/R SMP...

**Comment by microbuilder** [comment]
I really need an afternoon or a day to look at this in the detail it deserves. Sorry for being slow with a review. I appreciate the effort to sort this partitioning problème out, though, and I'll try to block the required time off this week.

**Comment by microbuilder** [comment]
> > ```
> > int arch_swap(unsigned int key)
> > {
> > 	/* store off key and return value */
> > 	_current->arch.basepri = key;
> > 	_current->arch.swap_return_value = _k_neg_eagain;
> > 
> > 	z_arm_cortex_r_svc();
> > 	irq_unlock(key);
> > 
> > 	/* Context switch is performed here. Returning implies the
> > 	 * thread has been context-switched-in again.
> > 	 */
> > 	return _current->arch.swap_return_value;
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This is the code from swap.c I was referring to. As far as I can tell, the _current->arch.basepri assignment would only work on a Cortex-M.
> 
> Or maybe... we might totally drop the `arch_swap` and use `USE_SWITCH` to support A/R to support A/R SMP...

@carlocaione You have more experience with SMP for Arm ... any thoughts here?

**Comment by carlocaione** [comment]
> @carlocaione You have more experience with SMP for Arm ... any thoughts here?

My thoughts here are that we should leave this discussion for another PR. This PR should be only dealing with separating the current code without any fix / rework. Anyway the @povergoing proposal look reasonable to me but that's an entirely different job and task :)

File: CODEOWNERS
Comment (microbuilder) [inline CODEOWNERS:22]: This should probably be updated to `@SgrrZhf @povergoing @microbuilder` Since Ioannis is no longer involved with the project.

**Review by microbuilder — CHANGES_REQUESTED** [review]
Not a proper review, just something I noticed taking a quick look just now.

File: CODEOWNERS
Comment (microbuilder) [reply CODEOWNERS:22]: Sorry, ignore this and lets make a separate PR since there are a lot of instances of this, and I agree with @carlocaione ... lets separate M and A/R here, and nothing else. I'll make a PR to cleanup `CODEOWNERS` later.

Comment (ibirnbaum) [reply CODEOWNERS:22]: Updating the CODEOWNERS might also be a chance to officially blame me for the MMU stuff for everyone to see...

**Comment by povergoing** [comment]
> @carlocaione You have more experience with SMP for A/R ... any thoughts here?

> My thoughts here are that we should leave this discussion for another PR. This PR should be only dealing with separating the current code without any fix / rework. Anyway the @povergoing proposal look reasonable to me but that's an entirely different job and task :)

Really apologize that I caused your misunderstanding. I do mean this is definitely another job should not be included in this PR. Actually, I just want to point out, that `arch_swap` is not a big deal for A/R as we might drop it in the future.

**Review by microbuilder — CHANGES_REQUESTED** [review]
First, thanks for all the work on this! It's a non-trivial change, but this makes the code much easier to read and maintain.

The separation looks good to me, although it's hard to review every line changed in detail with something this size.

Can you please also add some notes to the next release notes about these major changes, though? Something like the following (change as you feel appropriate):

```rst
* ARM

  * Architectural support for Arm Cortex-M has been separated from Arm
    Cortex-A and Cortex-R. This includes separate source modules to handle
    tasks like IRQ management, exception handling, thread handling and swap.
    For implementation details see :github:`60031`.
```

**Comment by SgrrZhf** [comment]
mps2_an385:tests/net/lib/lwm2m/content_json/net.lwm2m.content_json
mps2_an385:tests/net/lib/lwm2m/content_plain_text/net.lwm2m.content_plain_text

The failure of the two test cases above is because they need `CONFIG_NEWLIB_LIBC=y`, but after https://github.com/zephyrproject-rtos/zephyr/pull/57340, This test case will use `CONFIG_PICOLIBC=y` by default. So we need to create a new PR to fix it.

I wonder, is there any possibility of ignoring this failure and merging this PR?  😂

**Comment by microbuilder** [comment]
> I wonder, is there any possibility of ignoring this failure and merging this PR? 😂

Failures are unrelated, I agree, but lets wait until Monday to merge this in anyway, just because it's a non-trivial change and then we aren't stuck creating hotfix PRs Saturday at 23:00 just in case? 😅

**Comment by microbuilder** [comment]
@SgrrZhf Do you mind rebasing this for a clean CI run?

**Comment by SgrrZhf** [comment]
> @SgrrZhf Do you mind rebasing this for a clean CI run?

Done.

**Review by microbuilder — DISMISSED** [review]
Thanks for all the work on this.

**Review by povergoing — DISMISSED** [review]

**Review by carlocaione — DISMISSED** [review]

**Review by ibirnbaum — DISMISSED** [review]
Thank you for this major improvement!

**Comment by povergoing** [comment]
> Thanks for all the work on this.

Hi @microbuilder, shall we remove the `DNM` label? :)

**Comment by fabiobaltieri** [comment]
@SgrrZhf could you rebase please? I think it clashed with https://github.com/zephyrproject-rtos/zephyr/commit/1fd3171f7bb9dbc2733088861defb469c82c2110, couple header removals

**Comment by SgrrZhf** [comment]
> @SgrrZhf could you rebase please? I think it clashed with [1fd3171](https://github.com/zephyrproject-rtos/zephyr/commit/1fd3171f7bb9dbc2733088861defb469c82c2110), couple header removals

Done

**Review by carlocaione — APPROVED** [review]

**Review by fabiobaltieri — APPROVED** [review]

**Review by povergoing — APPROVED** [review]

**Review by microbuilder — APPROVED** [review]

**Comment by tejlmand** [comment]
this PR has broken armclang support.

When doing major rework, please remember to test with armclang, or ping someone who can carry out such test.

With armclang i'm now seeing errors like this:
```
CMake Error at /projects/github/ncs/zephyr/cmake/linker/armlink/target.cmake:74 (target_link_libraries):
  Error evaluating generator expression:                                                                                                                                                                                                      

    $<TARGET_OBJECTS:arch__arm__core__aarch32__cortex_m>                                                                                                                                                                                      

  Objects of target "arch__arm__core__aarch32__cortex_m" referenced but no                                                                                                                                                                    
  such target exists.                                                                                                                                                                                                                         
Call Stack (most recent call first):                                                                                                                                                                                                          
  /projects/github/ncs/zephyr/CMakeLists.txt:1427 (toolchain_ld_link_elf)        
```

**Comment by tejlmand** [comment]
@fabiobaltieri FYI

**Comment by fabiobaltieri** [comment]
> When doing major rework, please remember to test with armclang, or ping someone who can carry out such test.

Is there any fundamental blocker from having an armclang run in CI? Sounds like the only sustainable option to me.

**Comment by tejlmand** [comment]
>  We have a separate clang run already so unless there's some licensing issue I don't see why not.

we do have some licenses for armclang for develoment, but I guess no one has had the time to look into getting this up and running with CI.

Note also that we should checkup with arm that we are permitted to use those licenses in CI.

**Comment by tejlmand** [comment]
i'm curious why this PR, which touches so many files in 11 commits, only have a manifest label, and no `build-system` label or similar.

**Comment by tejlmand** [comment]
Fixed here: https://github.com/zephyrproject-rtos/zephyr/pull/62590

**Comment by fabiobaltieri** [comment]
> i'm curious why this PR, which touches so many files in 11 commits, only have a manifest label, and no `build-system` label or similar.

I don't think it touches any files under "Build system" (https://github.com/zephyrproject-rtos/zephyr/blob/main/MAINTAINERS.yml#L334-L342), but regardless, the script that does the tagging does not apply more than 10 tags, this check kicked in: https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/set_assignees.py#L158

**Comment by fabiobaltieri** [comment]
> When doing major rework, please remember to test with armclang, or ping someone who can carry out such test.

> Note also that we should checkup with arm that we are permitted to use those licenses in CI.

This is not really a sustainable solution at the change volume that the project is dealing with. It also does not seem fair to block against a proprietary system that only few developer have access to.

If the tool is proprietary and we are not allowed to run in CI then regressions will have to be detected and fixed out band.

**Pull request merged**
